// @harlem/extension-storage/dist/index.bundler.mjs
// src/constants.ts
var SENDER = "extension:storage";
var MUTATIONS = {
  sync: "extension:storage:sync"
};

// src/index.ts
import {
  EVENTS
} from "@harlem/core";
import {
  functionIdentity,
  matchGetFilter,
  objectFromPath,
  objectSet,
  objectTrace,
  typeIsNil
} from "@harlem/utilities";
function getOptions(options) {
  return {
    type: "local",
    prefix: "harlem",
    sync: true,
    restore: false,
    include: "*",
    exclude: [],
    serialiser: (state) => JSON.stringify(state),
    parser: (value) => JSON.parse(value),
    branch: functionIdentity,
    ...options
  };
}
function storageExtension(options) {
  const _options = getOptions(options);
  const {
    type,
    prefix,
    sync,
    restore,
    include,
    exclude,
    serialiser,
    parser,
    branch
  } = _options;
  return (store) => {
    if (store.flags.has("ssr:server")) {
      const noop = () => {
      };
      return {
        startStorageSync: noop,
        stopStorageSync: noop,
        clearStorage: noop,
        restoreStorage: noop
      };
    }
    store.register("extensions", "storage", () => _options);
    const {
      value,
      getNodes,
      resetNodes
    } = objectTrace();
    const storage = type === "session" ? sessionStorage : localStorage;
    const storageKey = prefix ? `${prefix}:${store.name}` : store.name;
    const mutationFilter = matchGetFilter({
      include,
      exclude
    });
    resetNodes();
    branch(value);
    function startStorageWrite() {
      store.on(EVENTS.mutation.success, (event) => {
        if (!event || event.data.name === MUTATIONS.sync || !mutationFilter(event.data.name)) {
          return;
        }
        try {
          const state = objectFromPath(store.state, getNodes());
          storage.setItem(storageKey, serialiser(state));
        } catch (e) {
          console.warn("Failed to write to storage");
        }
      });
    }
    function syncStorage(value2) {
      store.write(MUTATIONS.sync, SENDER, (state) => objectSet(state, getNodes(), parser(value2)));
    }
    function listener({ key, storageArea, newValue }) {
      if (storageArea === storage && key === storageKey && newValue) {
        syncStorage(newValue);
      }
    }
    function startStorageSync() {
      window.addEventListener("storage", listener);
    }
    function stopStorageSync() {
      window.removeEventListener("storage", listener);
    }
    function clearStorage() {
      storage.removeItem(storageKey);
    }
    function restoreStorage() {
      const value2 = storage.getItem(storageKey);
      if (!typeIsNil(value2)) {
        syncStorage(value2);
      }
    }
    store.once(EVENTS.store.created, () => {
      startStorageWrite();
      if (sync) {
        startStorageSync();
      }
      if (restore) {
        restoreStorage();
      }
    });
    store.once(EVENTS.store.destroyed, () => stopStorageSync());
    return {
      startStorageSync,
      stopStorageSync,
      clearStorage,
      restoreStorage
    };
  };
}
export {
  storageExtension as default
};
